packages/session-replay-browser/src/session-replay.ts

```diff
@@ -0,0 +1,530 @@
+import { getAnalyticsConnector, getGlobalScope } from '@amplitude/analytics-client-common';
+import { BaseTransport, Logger } from '@amplitude/analytics-core';
+import { Logger as ILogger, ServerZone, Status } from '@amplitude/analytics-types';
+import * as IDBKeyVal from 'idb-keyval';
+import { pack, record } from 'rrweb';
+import { SessionReplayConfig } from './config';
+import {
+  BLOCK_CLASS,
+  DEFAULT_SESSION_REPLAY_PROPERTY,
+  MASK_TEXT_CLASS,
+  MAX_EVENT_LIST_SIZE_IN_BYTES,
+  MAX_IDB_STORAGE_LENGTH,
+  MAX_INTERVAL,
+  MIN_INTERVAL,
+  SESSION_REPLAY_EU_URL as SESSION_REPLAY_EU_SERVER_URL,
+  SESSION_REPLAY_SERVER_URL,
+  STORAGE_PREFIX,
+  defaultSessionStore,
+} from './constants';
+import { isSessionInSample, maskInputFn } from './helpers';
+import {
+  MAX_RETRIES_EXCEEDED_MESSAGE,
+  MISSING_API_KEY_MESSAGE,
+  MISSING_DEVICE_ID_MESSAGE,
+  STORAGE_FAILURE,
+  UNEXPECTED_ERROR_MESSAGE,
+  getSuccessMessage,
+} from './messages';
+import {
+  AmplitudeSessionReplay,
+  Events,
+  IDBStore,
+  IDBStoreSession,
+  SessionReplayConfig as ISessionReplayConfig,
+  RecordingStatus,
+  SessionReplayContext,
+  SessionReplayOptions,
+} from './typings/session-replay';
+
+export class SessionReplay implements AmplitudeSessionReplay {
+  name = '@amplitude/session-replay-browser';
+  config: ISessionReplayConfig | undefined;
+  loggerProvider: ILogger;
+  storageKey = '';
+  retryTimeout = 1000;
+  events: Events = [];
+  currentSequenceId = 0;
+  private scheduled: ReturnType<typeof setTimeout> | null = null;
+  queue: SessionReplayContext[] = [];
+  stopRecordingEvents: ReturnType<typeof record> | null = null;
+  maxPersistedEventsSize = MAX_EVENT_LIST_SIZE_IN_BYTES;
+  interval = MIN_INTERVAL;
+  timeAtLastSend: number | null = null;
+
+  constructor() {
+    this.loggerProvider = new Logger();
+  }
+
+  async init(apiKey: string, options: SessionReplayOptions) {
+    this.config = new SessionReplayConfig(apiKey, options);
+    this.loggerProvider = this.config.loggerProvider;
+
+    this.loggerProvider.log('Installing @amplitude/session-replay-browser.');
+
+    this.storageKey = `${STORAGE_PREFIX}_${this.config.apiKey.substring(0, 10)}`;
+
+    const globalScope = getGlobalScope();
+    if (globalScope) {
+      globalScope.addEventListener('blur', this.blurListener);
+      globalScope.addEventListener('focus', this.focusListener);
+    }
+
+    if (globalScope && globalScope.document && globalScope.document.hasFocus()) {
+      await this.initialize(true);
+    }
+  }
+
+  setSessionId(sessionId: number) {
+    if (!this.config) {
+      this.loggerProvider.error('Session replay init has not been called, cannot set session id.');
+      return;
+    }
+    this.stopRecordingAndSendEvents(this.config.sessionId);
+    this.config.sessionId = sessionId;
+    this.events = [];
+    this.currentSequenceId = 0;
+    this.recordEvents();
+  }
+
+  getSessionRecordingProperties() {
```

Comment:
this interface implies that these properties are static. we probably want to be clear when the user should refetch these properties. I.e., a new session might be created and not sampled, opt-out might be triggered in the middle of a recording.

Reply:
oh actually, it looks like opting out in the middle of a session isn't supported, which I think is the right behavior.

Reply:
@jackson-amplitude take a look at the README and let me know if the way I've described it there provides that clarity! Or if you have a suggestion for a better name for this method very happy to hear it also